[KEP-3104] Update KEP with credential plugin allowlist

pmengelbert opened the pull request:
<!-- 
	Please use the following format when naming your PR
	< Issue Number >:< Issue Description >
	e.g. KEP-000: adding beta graduation criteria
	
	Avoid using phrases like `fixes #NNNN` in the description
	unless the pull request is to change the KEP status to 
	implemented or KEP has been deprecated.
-->

<!-- short description of work done in PR e.g. updating milestone, adding new KEP, adding test requirements… -->  
- One-line PR description: Update kuberc KEP with credential plugin allowlist

<!-- link to the k/enhancements issue -->
- Issue link: #3104 

<!-- other comments or additional information -->
- Other comments:
- 
Currently, the kubeconfig may specify arbitrary binaries to run as
client-go credential plugins. Due to the fact that kubeconfig files are
often generated, and because they contain a lot of noise, there is
significant friction in manually inspecting the kubeconfig for
suspicious binaries after it is generated.

To encourage secure behavior, we want to introduce an allowlist to the
kuberc, which will describe the conditions under which a binary plugin
may be run. Currently the only condition is the name (absolute path or
basename of a binary found in `PATH`).

enj commented:
/sig auth

enj commented:
This LGTM from a SIG Auth perspective (I added us as a participating SIG).

ardaguclu commented on keps/sig-cli/3104-introduce-kuberc/README.md:
```diff
@@ -331,6 +337,31 @@ intended behavior and realizing that targeting options effectively addresses the
+binary in question must meet all of the required conditions in that entry in
+order to be executed. At the outset, the entry object will have only one field,
+`name`. The path of the binary specified in the kubeconfig will be compared
+against that named in the `name` field. This field may contain a basename, or
```
Actually this depends on the behavior of `os.LookPath`

pmengelbert replied:
This reminds me: the function is `exec.LookPath`, not `os.LookPath`. I made a mistake in this PR and will update it accordingly.

`exec.LookPath` doesn't do tilde expansion, but that is POSIX shell behavior and not exec syscall behavior (at least on linux). I believe we should stop short of emulating POSIX substitution rules; the user can add e.g. `$HOME/bin` to their PATH if they want.

pmengelbert replied:
> This reminds me: the function is exec.LookPath, not os.LookPath. I made a mistake in this PR and will update it accordingly.

This has been fixed

ardaguclu commented on keps/sig-cli/3104-introduce-kuberc/README.md:
```diff
@@ -331,6 +337,31 @@ intended behavior and realizing that targeting options effectively addresses the
+execute.  If no criteria set succeeds after comparing the binary to all sets of
+criteria, the operation will be immediately aborted and an error returned. If
+`credentialPluginAllowlist` is not provided, or is explicitly made `nil`, all
+binaries will be allowed. If `credentialPluginAllowlist`'s value is set to the
```
I think, we should mention about this list will not be applied for some commands such as `kubectl config view` and the reason?.

ardaguclu commented on keps/sig-cli/3104-introduce-kuberc/README.md:
```diff
@@ -331,6 +337,31 @@ intended behavior and realizing that targeting options effectively addresses the
+criteria, the operation will be immediately aborted and an error returned. If
+`credentialPluginAllowlist` is not provided, or is explicitly made `nil`, all
+binaries will be allowed. If `credentialPluginAllowlist`'s value is set to the
+empty list `[]`, *all binaries will be prohibited*.
```
This is API behavior. So I wonder opinions from @liggitt. 
credentialPluginAllowlist: nil -- all allowed
credentialPluginAllowlist: [] -- all prohibited

pmengelbert replied:
I have pushed changes addressing this, opting for the flat structure.

pmengelbert commented on keps/sig-cli/3104-introduce-kuberc/README.md:
```diff
@@ -331,6 +337,31 @@ intended behavior and realizing that targeting options effectively addresses the
 use cases. During command execution, a merge will be occur, with inline overrides
 taking precedence over the defaults.
 
+`credentialPluginAllowlist` allows the end-user to provide an array of objects
```
There is an example [in this same PR, below](https://github.com/kubernetes/enhancements/pull/5479/files#diff-7feceb7c2590858d1c6eefb4997a71a675f52ffec384b759940378bff2a646daR390). Or are you asking for something different?

ardaguclu commented on keps/sig-cli/3104-introduce-kuberc/README.md:
```diff
@@ -331,6 +337,52 @@ intended behavior and realizing that targeting options effectively addresses the
+reason. The user should instead use `credentialPluginPolicy: DisableAll` in
+this case.
+
+Commands that don't create a client, such as `kubectl config view` will not be
```
is there any other command that will be exceptional like kubectl config view?

ardaguclu replied:
That sounds like a reasonable plan for me.

ardaguclu reviewed (commented):
Overall looks good to me. I'll defer the final approval from SIG-CLI POV to @soltysh

soltysh commented on keps/sig-cli/3104-introduce-kuberc/README.md:
```diff
@@ -331,6 +337,52 @@ intended behavior and realizing that targeting options effectively addresses the
+`EnableAll` is the default if no policy is provided, and will allow the
+execution of any plugin. `DisableAll` will not allow any credential plugins to
+be executed. The behavior of `Allowlist` is determined by the top-level
+`credentialPluginAllowlist` field, described below.
```
Sorry, I missed that :sweat:

soltysh commented on keps/sig-cli/3104-introduce-kuberc/README.md:
```diff
@@ -331,6 +337,52 @@ intended behavior and realizing that targeting options effectively addresses the
+binary in question must meet all of the required conditions in that entry in
+order to be executed (i.e. they are combined with logical AND).
+
+Each element in the allowlist is a set of criteria; if the binary in question
```
This one still holds, though.

BenTheElder reviewed (commented):
Do we think that processes with permission to mutate `KUBECONFIG` will commonly not have permission to edit kuberc? 

I suspect currently many many users are using the default kubeconfig and kuberc location which are in the same directory (even if they don't have to be and could have different permissions ...)

Seems like we might need to call out how to properly leverage this to not give a false sense of security?

pmengelbert commented:
Thanks to everyone for the feedback, and apologies for the delay. Before addressing comments I wanted to seek alignment between sig-cli, sig-auth, and sig-api-machinery since this feature will touch on areas governed by all three. I'll be addressing feedback ASAP and will get everything in good shape before Enhancements Freeze next week.

pmengelbert commented on keps/sig-cli/3104-introduce-kuberc/README.md:
```diff
@@ -319,6 +324,7 @@ fields (`apiVersion`, `kind`):
 
 * `aliases` Allows users to declare their own command aliases, including options and values.
 * `defaults` Enables users to set default options to be applied to commands.
+* `credentialPluginAllowlist` Enables users to specify criteria for trusting binaries to be executed by the client-go credential plugin system.
```
done

liggitt commented on keps/sig-cli/3104-introduce-kuberc/README.md:
```diff
@@ -331,6 +339,54 @@ intended behavior and realizing that targeting options effectively addresses the
+overall result of a check against the allowlist will be the logical OR of the
+individual checks against the allowlist's entries. Each allowlist entry MUST
+have at least one nonempty field describing the conditions required for the
+plugin's execution. If multiple fields are specified within an entry, the
```
the fields within a single item in the allowlist:

```
allowlist:
- name: x
- name: y,
- name: z
```

there's only one field now (name), but the structure allows for additional fields in the future

ardaguclu replied:
It makes sense now. Thanks.

liggitt commented:
go ahead and squash down commits and take out the `@mention` in the commit message tripping up the presubmit check

pmengelbert commented:
> go ahead and squash down commits and take out the `@mention` in the commit message tripping up the presubmit check

@liggitt  done

pmengelbert commented on keps/sig-cli/3104-introduce-kuberc/README.md:
```diff
@@ -331,6 +339,54 @@ intended behavior and realizing that targeting options effectively addresses the
+paths must match.
+
+If `credentialPluginPolicy` is set to `Allowlist`, but a `credentialPluginAllowlist`
+is not provided, it will be considered an error and the operation will fail.
```
done

pmengelbert commented on keps/sig-cli/3104-introduce-kuberc/README.md:
```diff
@@ -331,6 +337,52 @@ intended behavior and realizing that targeting options effectively addresses the
 use cases. During command execution, a merge will be occur, with inline overrides
 taking precedence over the defaults.
 
+`credentialPluginPolicy` determines the strategy by which credential plugins
```
done

pmengelbert commented on keps/sig-cli/3104-introduce-kuberc/README.md:
```diff
@@ -331,6 +337,52 @@ intended behavior and realizing that targeting options effectively addresses the
+returned.
+
+At the outset, the entry object will have only one field, `name`. The path of
+the binary specified in the kubeconfig will be compared against that named in
```
The table is done now as well

pmengelbert commented:
All comments have been addressed and this is ready for another round of review.

ardaguclu commented:
I'm approving from sig-cli POV
/approve
I'm also adding hold to give a chance to @soltysh for one last check
/hold
